### PR TITLE
(hactoberfest): Fix code scanning alert #3821

### PR DESCRIPTION
**Pull request by BLANKatGITHUB**
As first a student and first time contributor, I didn't understand fully what I had to do , but I tried my best. Thanks for giving me opportunity to learn :) .

### CODE REVIEW

**Comment by romange** [comment]
Thanks! A few questions:
1. Why github actions use hashes?
2. what step-security/harden-runner does? why do we need it in all the workflows?

File: .github/workflows/codeql.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        # CodeQL supports [ $supported-codeql-languages ]
+        # Learn more about CodeQL language support at https://aka.ms/codeql-docs/language-support
+
+    steps:
+      - name: Harden Runner
+        uses: step-security/harden-runner@91182cccc01eb5e619899d80e4e971d6181294a7 # v2.10.1
+        with:
+          egress-policy: audit
+
+      - name: Checkout repository
```
Comment (romange) [inline .github/workflows/codeql.yml:48]: you should checkout with submodules. See how we checkout in other places. 

what this action does?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -30,3 +30,23 @@ repos:
     rev: 23.7.0
     hooks:
       - id: black
+  - repo: https://github.com/gitleaks/gitleaks
```
Comment (romange) [inline .pre-commit-config.yaml:33]: what are these?

**Comment by romange** [comment]
My main comment is that this PR is huge and I prefer we take it by baby steps - since both you and I do not have much experience in this domain. Please choose 1-2 improvements you want to add and submit only them, no need to add everything at once.

**Comment by BLANKatGITHUB** [comment]
I appreciate your feedback, and I am sorry for inconvenience that I may have caused, I am somewhat familiar with GitHub Actions but never used them in real projects, I will try my best to solve these and add them incrementally . I hope you can guide me if I face a problem thanks.

**Comment by BLANKatGITHUB** [comment]
> Thanks! A few questions:
> 
> 1. Why github actions use hashes?

GitHub actions use hashes to use the specific commit or SHA of an action to ensure stability.

> 2. what step-security/harden-runner does? why do we need it in all the workflows?

 step-security/harden-runner provides some security features like network egress filtering and runtime security.

File: .github/workflows/codeql.yml (hunk 1 above)
Comment (BLANKatGITHUB) [reply .github/workflows/codeql.yml:48]: used to monitor and log all outgoing network traffic during your CI/CD workflows without blocking any traffic.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (BLANKatGITHUB) [reply .pre-commit-config.yaml:33]: gitleaks repo is used to detect and prevent hardcoded secrets like passwords, API keys and tokens in git repositries.

**Comment by romange** [comment]
I prefer we won't use hashes and not add `harden-runner` - this should reduce the diff significantly.

**Comment by BLANKatGITHUB** [comment]
I will create a new pull request requesting change in one workflow.

**Pull request closed without merging**